Add generate_series() udtf (and introduce 'lazy' `MemoryExec`)

2010YOUY01 opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes https://github.com/apache/datafusion/issues/10069

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
To make it easier to test memory-limited queries, i was looking for a datasource which can generate long streams, and itself doesn't consume lots of memory.
There are several existing approaches are close, but can't fully satisfy this requirement:
- `MemoryExec` have to buffer all output data up front
- `select * from unnest(generate_series(1,10000))` also have to buffer all output during construction
- It's possible to do such tests on large files, however it would be hard to setup and takes longer to run

So in this PR, a new `StreamingMemoryExec` is introduced, it can be used to generate batches lazily by defining a 'iterator' on `RecordBatch` in its interface. And a new user-defined table function is implemented with this new execution plan.

This function's behavior is kept consistent with [DuckDB's generate_series() table function](https://duckdb.org/docs/sql/functions/list#generate_series), and only 2 argument variant is implemented for now

I also think this UDTF can be useful to tests/micro-benchmarks in general

```
> select * from generate_series(1,3);
+-------+
| value |
+-------+
| 1     |
| 2     |
| 3     |
+-------+
```

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
1. Move `TableFunctionImpl` from `datafusion` crate -> `datafusion-catalog` crate, so new UDTFs defined in `datafusion-functions-table` crate won't have circular dependency with `datafusion` crate (`TableFunctionImpl` depends on `TableProvider` in `datafusion-catalog` crate, so I think it's the best place to move it to)
2. Introduced `StreamingMemoryExec`
3. Implemented a new UDTF `generate_series()`

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
4. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->
Unit tests for `StreamingMemoryExec`, and `sqllogictest`s for `generate_series()` UDTF

## Are there any user-facing changes?

No
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

2010YOUY01 commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
+    ) -> Result<Self> {
+        let cache = PlanProperties::new(
+            EquivalenceProperties::new(Arc::clone(&schema)),
+            Partitioning::RoundRobinBatch(generators.len()),
```
Is this field mean: Let's say a exec only have 1 generator, the optimizer will try to insert a `RepartitionExec` to the output of `StreamingMemoryExec`, and use round robin to repartition output batches to `target_partitions` number?
I found it works on some aggregate queries, but not for a sort query
```
> explain select * from generate_series(1, 10000) as t1(v1) order by v1;
+---------------+----------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                           |
+---------------+----------------------------------------------------------------------------------------------------------------+
| logical_plan  | Sort: t1.v1 ASC NULLS LAST                                                                                     |
|               |   SubqueryAlias: t1                                                                                            |
|               |     Projection: tmp_table.value AS v1                                                                          |
|               |       TableScan: tmp_table projection=[value]                                                                  |
| physical_plan | SortExec: expr=[v1@0 ASC NULLS LAST], preserve_partitioning=[false]                                            |
|               |   ProjectionExec: expr=[value@0 as v1]                                                                         |
|               |     StreamingMemoryExec: partitions=1, batch_generators=[generate_series: start=1, end=10000, batch_size=8192] |
|               |                                                                                                                |
+---------------+----------------------------------------------------------------------------------------------------------------+
```
I want to know is it correct here, and this sort query in theory can be made parallelized by changing somewhere in optimization phase? 🤔

alamb replied:
Maybe you can look at https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.repartitioned or similar methods 🤔

jayzhan211 replied:
I'm not sure about this too

jayzhan211 commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
+/// Stream that generates record batches on demand
+pub struct StreamingMemoryStream {
+    schema: SchemaRef,
+    generator: Box<dyn StreamingBatchGenerator>,
```
Should we use `Arc` given it makes sense to have generator generate across threads

jayzhan211 commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
+    /// Schema representing the data
+    schema: SchemaRef,
+    /// Functions to generate batches for each partition
+    batch_generators: Vec<Box<dyn StreamingBatchGenerator>>,
```
```suggestion
    batch_generators: Vec<Arc<dyn StreamingBatchGenerator>>,
```

jayzhan211 commented on datafusion/catalog/src/table.rs:
```diff
@@ -297,3 +299,40 @@ pub trait TableProviderFactory: Debug + Sync + Send {
 }
+
+/// A trait for table function implementations
+pub trait TableFunctionImpl: Debug + Sync + Send {
```
```suggestion
/// Returns this function's name
    fn name(&self) -> &str;
```

This is much more consistent with other UDFs

2010YOUY01 replied:
This is the part of the refactor, changing this requires several fixes. We can leave it to a separate PR to make this PR smaller

berkaysynnada replied:
I also recommend to take `name()` inside `TableFunctionImpl`

alamb replied:
I think doing it as a follow on PR makes sense to me -- perhaps we can file a ticket

Another thing that would be nice for a follow on PR is to move this trait into its own module (`catalog/src/table_function.rs` perhaps)

2010YOUY01 replied:
Filed https://github.com/apache/datafusion/issues/13613

jayzhan211 commented on datafusion/sqllogictest/test_files/table_functions.slt:
```diff
@@ -0,0 +1,112 @@
+SELECT * FROM generate_series(5, 1)
+
+statement error DataFusion error: Error during planning: First argument must be an integer literal
+SELECT * FROM generate_series(NULL, 5)
```
We might not return error for this query
```
D SELECT * FROM generate_series(null, 5);
┌─────────────────┐
│ generate_series │
│      int64      │
├─────────────────┤
│     0 rows      │
└─────────────────┘
```

2010YOUY01 replied:
Agreed, it's a good idea to align the behavior with DuckDB

jayzhan211 commented on datafusion/sqllogictest/test_files/table_functions.slt:
```diff
@@ -0,0 +1,112 @@
+SELECT * FROM generate_series(NULL, 5)
+
+statement error DataFusion error: Error during planning: Second argument must be an integer literal
+SELECT * FROM generate_series(1, NULL)
```
```
D SELECT * FROM generate_series(1, null);
┌─────────────────┐
│ generate_series │
│      int64      │
├─────────────────┤
│     0 rows      │
└─────────────────┘
```

jayzhan211 commented on datafusion/sqllogictest/test_files/table_functions.slt:
```diff
@@ -0,0 +1,112 @@
+statement error DataFusion error: Error during planning: Second argument must be an integer literal
+SELECT * FROM generate_series(1, NULL)
+
+statement error DataFusion error: Error during planning: generate_series expects 2 arguments
```
```
D SELECT * FROM generate_series(1, 5, null);
┌─────────────────┐
│ generate_series │
│      int64      │
├─────────────────┤
│     0 rows      │
└─────────────────┘
```

2010YOUY01 replied:
I'll open an issue for 3-arg support later

2010YOUY01 commented:
Thank you for the comments @jayzhan211 , I have updated.

Now I think the concurrent generator might not be very straightforward, I'll first leave some rationale here.
If we can come to agreement I'll add more doc to it (else revert back to `Box` version for simplicity)

----
### Rationale for `StreamingMemoryStream` design
Suppose we want to run `select ... from generate_series(1,100)` in two partitions
And the underlying batch generator is wrapped with Mutex
```
pub struct StreamingMemoryStream {
    ...
    generator: Arc<Mutex<dyn StreamingBatchGenerator>>,
}
```
It's possible to implement the UDTF in 3 ways:
1.
```
[generator_1_50]   --- [StreamingMemoryStream  stream1] --> xxStream1
[generator_50_100] --- [StreamingMemoryStream  stream2] --> xxStream1
```
2.
```
[generator_1_100] --- [StreamingMemoryStream  stream1] --> Repartition --> xxStream1
                                                                       |-> xxStream2                                                                                
```
3.
```
[generator_1_100] --- [StreamingMemoryStream  stream1] --> xxStream1
                  |-- [StreamingMemoryStream  stream2] --> xxStream2                                                                                                                                      
```
1 and 2 is the common pattern for datafusion scanning operators to do plan-time parallelism, `Mutex` is not used in this case
3 make the `StreamingBatchGenerator` being able to concurrently accessed by multiple streams. It's trying to make the interface more general-purpose for future usage, and won't be used by `generate_series()`

jayzhan211 commented:
> Thank you for the comments @jayzhan211 , I have updated.
> 
> Now I think the concurrent generator might not be very straightforward, I'll first leave some rationale here. If we can come to agreement I'll add more doc to it (else revert back to `Box` version for simplicity)
> 
> ### Rationale for `StreamingMemoryStream` design
> Suppose we want to run `select ... from generate_series(1,100)` in two partitions And the underlying batch generator is wrapped with Mutex
> 
> ```
> pub struct StreamingMemoryStream {
>     ...
>     generator: Arc<Mutex<dyn StreamingBatchGenerator>>,
> }
> ```
> 
> It's possible to implement the UDTF in 3 ways: 1.
> 
> ```
> [generator_1_50]   --- [StreamingMemoryStream  stream1] --> xxStream1
> [generator_50_100] --- [StreamingMemoryStream  stream2] --> xxStream1
> ```
> 
> 
> ```
> [generator_1_100] --- [StreamingMemoryStream  stream1] --> Repartition --> xxStream1
>                                                                        |-> xxStream2                                                                                
> ```
> 
> 
> ```
> [generator_1_100] --- [StreamingMemoryStream  stream1] --> xxStream1
>                   |-- [StreamingMemoryStream  stream2] --> xxStream2                                                                                                                                      
> ```
> 
> 1 and 2 is the common pattern for datafusion scanning operators to do plan-time parallelism, `generator` won't be accessed by multiple threads thus `Mutex` is redundant 3 make the `StreamingBatchGenerator` being able to concurrently accessed by multiple streams.
> 
> The `Mutex` is added to make it possible for case 3 (so the interface can be more general-purpose for future use cases)

Given the quick look, I'm not sure whether we need 3rd case. It seems the 1st case runs execution parallelly too. I would need to think about the advantage of the 3rd case over 1st case.

For 3rd case, if we need it, we might use `Arc<RwLock<dyn T>>` if `generate_next_batch` takes `&self`. It might be more efficiently.

berkaysynnada commented on datafusion/core/src/execution/session_state_defaults.rs:
```diff
@@ -119,6 +120,11 @@ impl SessionStateDefaults {
+    pub fn default_table_functions() -> Vec<Arc<TableFunction>> {
+        functions_table::all_default_table_functions()
+    }
+
```
Can we get the defaults from a singleton ? Like other default getters, with get_or_init or smth similar

2010YOUY01 replied:
This is a good idea, updated

berkaysynnada commented on datafusion/functions-table/src/generate_series.rs:
```diff
@@ -0,0 +1,180 @@
+#[derive(Debug, Clone)]
+struct GenerateSeriesState {
+    schema: SchemaRef,
+    _start: i64,
```
Is there any reason to keep this?

2010YOUY01 replied:
This field is for display, i added a comment to explain

berkaysynnada commented on datafusion/functions-table/src/generate_series.rs:
```diff
@@ -0,0 +1,180 @@
+    // Check input `exprs` type and number. Input validity check (e.g. start <= end)
+    // will be performed in `TableProvider::scan`
+    fn call(&self, exprs: &[Expr]) -> Result<Arc<dyn TableProvider>> {
+        // TODO: support 3 arguments following DuckDB:
```
There is also a one length argument usage, perhaps can be added as todo

berkaysynnada commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+pub trait StreamingBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
```
I am not very satisfied with these `Streaming... `naming. Can we describe the behavior better with a different naming for these? Perhaps we could use 'lazy' term

berkaysynnada reviewed (approved):
Thank you @2010YOUY01. I reviewed the changes and LGTM. I have a few minor comments and one question: I noticed another approach of `generate_series()`, which can be used like this: 
`SELECT generate_series(1, 5)` 
I assume it is not a udtf in that context. Does this implementation leave room for that usage?

2010YOUY01 commented:
> Thank you @2010YOUY01. I reviewed the changes and LGTM. I have a few minor comments and one question: I noticed another approach of `generate_series()`, which can be used like this: `SELECT generate_series(1, 5)` I assume it is not a udtf in that context. Does this implementation leave room for that usage?

Thank you for the review @berkaysynnada 

The answer to the question is yes, they are different function with the same name.
Using same name is probably not the best idea, but we do this to keep the behavior the same as DuckDB.
I added a sql test to make sure they can work together (like `select generate_series() from generate_series()`)

----
@jayzhan211 I thought about the sharing generator issue again, I think making a interface more flexible is important, so I kept the lock, and updated `RwLock` as you suggested

alamb commented on Cargo.toml:
```diff
@@ -30,6 +30,7 @@ members = [
     "datafusion/functions",
     "datafusion/functions-aggregate",
     "datafusion/functions-aggregate-common",
+    "datafusion/functions-table",
```
100%

alamb commented on datafusion/functions-table/src/generate_series.rs:
```diff
@@ -0,0 +1,180 @@
+#[derive(Debug, Clone)]
+struct GenerateSeriesState {
+    schema: SchemaRef,
+    _start: i64, // Kept for display
```
does it need to be prefixed with `_` if it is used for display 🤔

alamb commented on datafusion/functions-table/src/generate_series.rs:
```diff
@@ -0,0 +1,180 @@
+    // Check input `exprs` type and number. Input validity check (e.g. start <= end)
+    // will be performed in `TableProvider::scan`
+    fn call(&self, exprs: &[Expr]) -> Result<Arc<dyn TableProvider>> {
+        // TODO: support 1 or 3 arguments following DuckDB:
```
can we possibly file a ticket to track this?

2010YOUY01 replied:
Tracked in https://github.com/apache/datafusion/issues/13615

alamb commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
+    schema: SchemaRef,
+    /// Functions to generate batches for each partition
+    batch_generators: Vec<Arc<RwLock<dyn LazyBatchGenerator>>>,
+    /// Total number of rows to generate for statistics
```
this comment doesn't seem quite right

alamb reviewed (approved):
Thank ou @2010YOUY01 @jayzhan211 and @berkaysynnada 

I left some structural comments / thoughts about how we might be able to make LazyMemoryExec more general (`StreamExec` or `StreamWrapperExec`) but in general I think this is a really nice first step and we could do additional refactorings as follow on PRs.

2010YOUY01 commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+pub trait LazyBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
```
This is a really good idea, filed https://github.com/apache/datafusion/issues/13614

2010YOUY01 commented:
Thank you @alamb , I have updated minor renaming/doc fixes in the latest commit, and opened issues for all other bigger follow on tasks

alamb commented:
Looks good -- let's get this merged 🚀
